scvi/models/_base.py

```diff
@@ -0,0 +1,792 @@
+import pandas as pd
+import numpy as np
+import inspect
+
+from anndata import AnnData
+from functools import partial
+from scvi.core._distributions import NegativeBinomial, ZeroInflatedNegativeBinomial
+from scvi.models._differential import DifferentialComputation
+from scvi.models._utils import (
+    scrna_raw_counts_properties,
+    _get_var_names_from_setup_anndata,
+)
+from scvi.dataset._utils import (
+    _check_anndata_setup_equivalence,
+    _check_nonnegative_integers,
+)
+from scvi import _CONSTANTS
+from typing import Optional, Union, List, Dict
+from scvi._compat import Literal
+from scvi.core.trainers import UnsupervisedTrainer
+from abc import ABC, abstractmethod
+from scvi.dataset import get_from_registry, transfer_anndata_setup
+
+logger = logging.getLogger(__name__)
+
+
+class VAEMixin:
+    def train(
+        self,
+        n_epochs=400,
+        train_size=0.9,
+        test_size=None,
+        lr=1e-3,
+        n_iter_kl_warmup=None,
+        n_epochs_kl_warmup=400,
+        frequency=None,
+        train_fun_kwargs={},
+        **kwargs
+    ):
+        # TODO figure out how to train after loading
+        # might need to have function which saved unsupervisedTrainer
+        if self.is_trained is False:
+            self.trainer = UnsupervisedTrainer(
+                self.model,
+                self.adata,
+                train_size=train_size,
+                test_size=test_size,
+                n_iter_kl_warmup=n_iter_kl_warmup,
+                n_epochs_kl_warmup=n_epochs_kl_warmup,
+                frequency=frequency,
+                use_cuda=self.use_cuda,
+                **kwargs,
+            )
+            self.train_indices = self.trainer.train_set.indices
+            self.test_indices = self.trainer.test_set.indices
+            self.validation_indices = self.trainer.validation_set.indices
+        # for autotune
+        if "n_epochs" not in train_fun_kwargs:
+            train_fun_kwargs["n_epochs"] = n_epochs
+        if "lr" not in train_fun_kwargs:
+            train_fun_kwargs["lr"] = lr
+        self.trainer.train(**train_fun_kwargs)
+        self.is_trained = True
+
+    @torch.no_grad()
+    def get_elbo(self, adata=None, indices=None, batch_size=128):
+
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
+
+        return -post.elbo()
+
+    @torch.no_grad()
+    def get_marginal_ll(
+        self, adata=None, indices=None, n_mc_samples=1000, batch_size=128
+    ):
+
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
+
+        return -post.marginal_ll(n_mc_samples=n_mc_samples)
+
+    @torch.no_grad()
+    def get_reconstruction_error(self, adata=None, indices=None, batch_size=128):
+
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
```

Comment:
Real question. How much overhead do we deal with when we create a new posterior?